lib/collection/src/shard/replica_set.rs

```diff
@@ -75,52 +84,200 @@ impl ReplicaSet {
+        while let Some(result) = futures.next().await {
+            if let ok @ Ok(_) = result {
+                return ok;
+            }
+        }
+        // at this point `captured_error` must be defined by construction
+        captured_error.unwrap()
+    }
 }
 
 #[async_trait::async_trait]
 impl ShardOperation for ReplicaSet {
     async fn update(
         &self,
-        _operation: CollectionUpdateOperations,
-        _wait: bool,
+        operation: CollectionUpdateOperations,
+        wait: bool,
     ) -> CollectionResult<UpdateResult> {
-        todo!()
+        // target all remote peers that are active
+        let active_remote_shards: Vec<_> = self
+            .remotes
+            .iter()
+            .filter(|rs| self.peer_is_active(&rs.peer_id))
+            .collect();
+
+        // local is defined AND the peer itself is active
+        let local_is_active = self.local.is_some() && self.peer_is_active(&self.this_peer_id);
+
+        if active_remote_shards.is_empty() && !local_is_active {
+            return Err(CollectionError::service_error(format!(
+                "The replica set for shard {} on peer {} has no active replica",
+                self.shard_id, self.this_peer_id
+            )));
+        }
+
+        let mut remote_futures = Vec::new();
+        for remote in active_remote_shards {
+            let op = operation.clone();
+            remote_futures.push(async move {
+                remote
+                    .update(op, wait)
+                    .await
+                    .map_err(|err| (remote.peer_id, err))
+            });
+        }
+
+        let all_res = match &self.local {
+            Some(local) if self.peer_is_active(&self.this_peer_id) => {
+                let local_update = async move {
+                    local
+                        .update(operation.clone(), wait)
+                        .await
+                        .map_err(|err| (self.this_peer_id, err))
+                };
+                let remote_updates = try_join_all(remote_futures);
+
+                // run local and remote shards read concurrently
+                try_join(remote_updates, local_update)
+                    .await
+                    .map(|(remote_res, _local_res)| remote_res)
+            }
+            _ => try_join_all(remote_futures).await,
+        };
+
+        match all_res {
+            Ok(results) => {
+                // return first result
+                match results.into_iter().next() {
+                    None => Err(CollectionError::service_error(format!(
+                        "None of the replicas replied for Replica set {} on peer {}",
+                        self.shard_id, self.this_peer_id
+                    ))),
+                    Some(res) => Ok(res),
+                }
+            }
+            Err((peer_id, err)) => {
+                // report failing `peer_id`
+                self.set_active(&peer_id, false)?;
```

Comment:
All peers need to sync the active state of replicas. So we can't use this shortcut here, instead submit a `SetShardReplicaState` operation to consensus

Reply:
Btw if it is the only place that requires `mut` state, then we can safely remove RwLock

Reply:
This is indeed the only place.

This means I need to submit the operation via the `Dispatcher` from the `ReplicaSet`?

However, the `Dispatcher` type lives in `Storage` and cannot be accessed from the `Collection` crate.
Is there another way?

Reply:
The callback cannot be passed via the `ShardOperation` without messing up the trait interface.

Therefore I spent a long time trying to have the callback being saved at the level of the `ReplicaSet` struct but I am facing an issue.

I am getting the following error when trying to use the callback.

```
error[E0618]: expected function, found `(dyn std::ops::Fn(u64) -> (dyn futures::Future<Output = ()> + 'static) + std::marker::Send + std::marker::Sync + 'static)`
  --> lib/collection/src/shard/replica_set.rs:41:9
   |
41 |         (*self.notify_peer_failure_cb)(peer_id)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^---------
   |         |
   |         call expression requires function

```

I need your Rust superpowers :pray:

Reply:
I pushed the changes that should fix the problem. They are not full - I left some stubs for the tests that we need to fix before we can merge it. But now it type-checks!

Reply:
It is just that under the hood async functions are built on top of pinned futures, I agree, that's not a very intuitive idea

Reply:
I have cleaned up the stubs to please `clippy` and get green tests in https://github.com/qdrant/qdrant/pull/997/commits/323b6055f56dd31afca04dabf3d7fdf2e31104bb
I also introduced a dummy callback to restore the collection's tests.
WYDYT?

Reply:
I think there is still 1 minor todo left to fix, but otherwise looks good!